hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java

```diff
@@ -0,0 +1,214 @@
+            int totalInvocations = log.getTotalInvocations();
+            if (totalInvocations == 1) {
+                LOGGER.warning(format("Slow operation detected: %s%n%s", currentOperation, log.getStackTrace()));
+            } else {
+                LOGGER.warning(format("Slow operation detected: %s (%d invocations)%n%s", currentOperation, totalInvocations,
+                        (totalInvocations % FULL_LOG_FREQUENCY == 0) ? log.getStackTrace() : log.getShortStackTrace()));
+            }
+        }
+    }
+
+    private SlowOperationLog getOrCreateSlowOperationLog(OperationRunner operationRunner, final Object operation) {
+        String prefix = "";
+        for (StackTraceElement stackTraceElement : operationRunner.currentThread().getStackTrace()) {
+            stackTraceStringBuilder.append(prefix).append(stackTraceElement.toString());
+            prefix = "\n";
+        }
+
+        final String stackTraceString = stackTraceStringBuilder.toString();
+        stackTraceStringBuilder.setLength(0);
+
+        return ConcurrencyUtil.getOrPutIfAbsent(slowOperationLogs, stackTraceString.hashCode(),
+                new ConstructorFunction<Integer, SlowOperationLog>() {
+                    @Override
+                    public SlowOperationLog createNew(Integer key) {
+                        return new SlowOperationLog(stackTraceString, operation);
+                    }
+                });
+    }
+
+    private boolean purge(long nowNanos, long lastLogPurge) {
+        if (nowNanos - lastLogPurge > logPurgeIntervalNanos) {
+            for (SlowOperationLog log : slowOperationLogs.values()) {
+                log.purgeInvocations(nowNanos, logRetentionNanos);
+                if (log.isEmpty()) {
+                    slowOperationLogs.remove(log.getStackTrace().hashCode());
+                }
+            }
+            return true;
+        }
+        return false;
+    }
+
+    private static class CurrentOperationData {
+        int operationHashCode;
+        long startNanos;
+        SlowOperationLog.Invocation invocation;
+    }
+
+    private class Scanner extends Thread {
+
+        public Scanner(HazelcastThreadGroup hazelcastThreadGroup) {
+            super(hazelcastThreadGroup.getThreadNamePrefix("SlowOperationDetectorThread"));
+        }
+
+        public void run() {
+            while (running) {
+                long lastLogPurge = System.nanoTime();
+
+                while (running) {
+                    long nowNanos = System.nanoTime();
+                    long nowMillis = System.currentTimeMillis();
+
+                    scan(nowNanos, nowMillis, genericOperationRunners, genericCurrentOperationData);
+                    scan(nowNanos, nowMillis, partitionOperationRunners, partitionCurrentOperationData);
+
+                    if (purge(nowNanos, lastLogPurge)) {
+                        lastLogPurge = nowNanos;
+                    }
+
```

Comment:
I would pull this into a method in the Scanner to keep the method nice and snappy.